packages/primitives-core/src/types.ts

```diff
@@ -1,10 +1,24 @@
-type NamedStyles<T> = { [P in keyof T]: unknown }
+// This is based on the StyleSheet types from @types/react-native
+
+type Falsy = undefined | null | false
+
+interface RecursiveArray<T>
+  extends Array<T | ReadonlyArray<T> | RecursiveArray<T>> {}
+
+type RegisteredStyle<T> = number & { __registeredStyleBrand: T }
+
+type StyleProp<T> =
+  | T
+  | RegisteredStyle<T>
+  | RecursiveArray<T | RegisteredStyle<T> | Falsy>
+  | Falsy
+
+type NamedStyles<T> = { [P in keyof T]: any }
 
-// This is based on the StyleSheet type from @types/react-native
 export interface AbstractStyleSheet {
   create<T extends NamedStyles<T> | NamedStyles<any>>(
     styles: T | NamedStyles<T>
-  ): T
+  ): { [P in keyof T]: number }
 
-  flatten(style?: unknown[]): unknown
+  flatten<T>(style?: StyleProp<T>): T extends (infer U)[] ? U : T
```

Comment:
Conceptually this might not be enough because `flatten` accepts nested arrays, so this should be recursive.

I wonder though if this is even needed. It feels a little bit like maybe some kind of a style prop should be generic. 

We already have not quite an accurate return type for the create method - in React Native `Stylesheet.create` returns the input object, it's basically an identity function:
https://github.com/facebook/react-native/blob/ba383baf357bc6def94280d28f69efbab341fe36/Libraries/StyleSheet/StyleSheet.js#L371

What I'm getting at is that we personally don't quite care about all of those types internally - we should just treat them as close to `unknown` as possible. We only need to make sure that given methods are available on the input `StyleSheet` in places like here:
https://github.com/emotion-js/emotion/blob/993557dcc253488eabf4673696c87085ded40772/packages/native/src/base.js#L7

Maybe we should just make a `StyleSheet` itself a generic? I'm not entirely sure about this but perhaps the best way to figure this out would be to convert this one together with `@emotion/native`. I feel like it's hard to figure out those types in isolation right now because the final design needs to be dictated by the needs of both `@emotion/primitives` and `@emotion/native`.

An offtopic note - I wish we didn't have those libraries at all 😅

Reply:
> We already have not quite an accurate return type for the create method - in React Native Stylesheet.create returns the input object, it's basically an identity function:

I could have sworn I tested `StyleSheet.create` and it returned an object whose values were numbers... I think they might have changed that in a semi-recent version of React Native. Anyway, I have corrected the `create` return type so that it matches `@types/react-native`.

> I feel like it's hard to figure out those types in isolation right now because the final design needs to be dictated by the needs of both @emotion/primitives and @emotion/native.

Yes, it is hard to figure them out in isolation. So do you think I should convert `@emotion/native` in this PR?

> An offtopic note - I wish we didn't have those libraries at all 😅

Totally agree LOL, I don't see how Emotion would be beneficial when using React Native...